cls_module/cls_module/memory/stm/aha.py

```diff
@@ -458,17 +478,44 @@ def forward_memory(self, inputs, targets, labels):
 
     outputs['pc'] = self.forward_pc(inputs=pc_cue)
 
-    losses['pm'], outputs['pm'] = self.forward_ae(name='pm', inputs=outputs['pc'], targets=targets)
-    losses['pm_ec'], outputs['pm_ec'] = self.forward_ae(name='pm_ec', inputs=outputs['pc'], targets=targets)
+    if self.config.get('msp_type', None) == 'ca1':
+      # During study, EC will drive the CA1
+      losses['ca1'], outputs['ca1'] = self.forward_ae(name='ca1', inputs=inputs, targets=inputs)
+      ca3_ca1_target = outputs['ca1']['encoding']
+
+      losses['ca3_ca1'], outputs['ca3_ca1'] = self.forward_ae(name='ca3_ca1',
```

Comment:
This is NOT within a conditional to ensure that it is in 'study'.
When in 'recall', this CA1 forward pass will be redundant.

Also, I thought it should fully train CA1 on all the current stimuli, and then train CA3-CA1. Is that how this will work? i.e. accumulate gradients across the batch, update weights, then we have a stable CA1 encoding to learn to reproduce by CA3-CA1. When hebbian, interleaved like this will work, because they'll both converge. But with batch training, I'm not sure.

Reply:
Yes, the forward pass will be redundant with regards to the final output as CA1 will be driven by CA3. I kept it because it can provide a target for CA3 during recall to provide a loss, otherwise we would have no target during recall to compare to.

Maybe the 'forward' naming is misleading, but each of our components' "forward_XX" method does both a forward pass and a backward pass (if training flag is on).

So in this case, the `forward_ae` with `ca1` has done a complete optimization step on the current batch, before moving onto the forward_ae with `ca3_ca1`